flowjax/bijections/planar.py

```diff
@@ -116,15 +153,39 @@ def get_act_scale(self):
         See appendix A1 in https://arxiv.org/pdf/1505.05770.pdf.
         """
         wtu = self._act_scale @ self.weight
-        m_wtu = -1 + jnp.log(1 + softplus(wtu))
+        m_wtu = -1 + jnp.log(1 + nn.softplus(wtu))
         return self._act_scale + (m_wtu - wtu) * self.weight / norm(self.weight) ** 2
 
     def inverse(self, y, condition=None):
-        raise NotImplementedError(
-            "The inverse planar transformation is not implemented.",
-        )
+        if self.activation != "leaky_relu":
+            raise NotImplementedError(
+                "The inverse planar transformation is only implemented with the leaky "
+                "relu activation function.",
+            )
+        return self.inverse_and_log_det(y, condition)[0]
 
     def inverse_and_log_det(self, y, condition=None):
-        raise NotImplementedError(
-            "The inverse planar transformation is not implemented.",
-        )
+        if self.activation != "leaky_relu":
+            raise NotImplementedError(
+                "The inverse planar transformation is only implemented with the leaky "
+                "relu activation function.",
+            )
+        # Expanding explanation as the inverse is not in the original paper.
+        # The derivation steps for the inversion are:
+        # 1. Let z = w^Tx+b
+        # 2. We want x=y-uσ(z), where σ is the leaky relu function.
+        # 3. Sub x=y-uσ(z) into z = w^Tx+b,
+        # 4. Solve for z, which gives z = (w^Ty+b)/(1+w^Tus), where s is the slope
+        #   σ'(z), i.e. s=1 if z>=0 and s=negative_slope otherwise. To find the
+        #   slope, it is sufficient to check the sign of the numerator w^Ty+b, rather
+        #   than z, as the denominator is constrained to be positive.
+        # 5. Compute inverse using x=y-uσ(z)
+
+        numerator = self.weight @ y + self.bias
+        relu_slope = jnp.where(numerator < 0, self.negative_slope, 1)
+        us = self.get_act_scale() * relu_slope
+        denominator = 1 + self.weight @ us
+        psi = relu_slope * self.weight
+        log_det = -jnp.log(jnp.abs(1 + us @ psi))
```

Comment:
Can you please double check if this is right? Inside the absolute value it is supposed to be 1 + us @ self.weight instead of 1 + us @ psi? I'm looking at Eq 12 of Rezende et al (2015).

Reply:
Thanks for spotting that. I was halfway between ``us @ weight`` and ``u @ psi``...